operators/pkg/controller/elasticsearch/driver/default.go

```diff
@@ -568,6 +512,161 @@ func removePodFromList(pods []corev1.Pod, pod corev1.Pod) []corev1.Pod {
+	// The goal is to:
+	// 1. scale sset up (eg. go from 3 to 5 replicas).
+	// 2. apply configuration changes on the sset resource, to be used for future pods creation/recreation,
+	//    but do not rotate pods yet.
+	// 3. do **not** apply replicas scale down, otherwise nodes would be deleted before
+	//    we handle a clean deletion.
+	for _, nodeSpec := range nodeSpecResources {
+		// always reconcile config (will apply to new & recreated pods)
+		if err := settings.ReconcileConfig(d.Client, es, nodeSpec.StatefulSet.Name, nodeSpec.Config); err != nil {
+			return err
+		}
+		if _, err := common.ReconcileService(d.Client, d.Scheme, &nodeSpec.HeadlessService, &es); err != nil {
+			return err
+		}
+		ssetToApply := nodeSpec.StatefulSet.DeepCopy()
+		actual, exists := actualStatefulSets.GetByName(ssetToApply.Name)
+		if exists && *ssetToApply.Spec.Replicas < *actual.Spec.Replicas {
+			// sset needs to be scaled down
+			// update the sset to use the new spec but don't scale replicas down for now
+			ssetToApply.Spec.Replicas = actual.Spec.Replicas
+		}
+		if err := sset.ReconcileStatefulSet(d.Client, d.Scheme, es, *ssetToApply); err != nil {
+			return err
+		}
+	}
+
+	// Phase 2: handle sset scale down.
+	// We want to safely remove nodes from the cluster, either because the sset requires less replicas,
+	// or because it should be removed entirely.
+	for i, actual := range actualStatefulSets {
+		expected, shouldExist := nodeSpecResources.StatefulSets().GetByName(actual.Name)
+		switch {
+		// stateful set removal
+		case !shouldExist:
+			target := 0
+			if err := d.scaleStatefulSetDown(results, &actualStatefulSets[i], target, esClient, observedState); err != nil {
+				return err
+			}
+		// stateful set downscale
+		case actual.Spec.Replicas != nil && *expected.Spec.Replicas < *actual.Spec.Replicas:
+			target := int(*expected.Spec.Replicas)
+			if err := d.scaleStatefulSetDown(results, &actualStatefulSets[i], target, esClient, observedState); err != nil {
+				return err
+			}
+		}
+	}
+
+	// TODO:
+	//  - safe node upgrade (rollingUpdate.Partition + shards allocation)
+	//  - change budget
+	//  - zen1, zen2
+	return nil
+}
+
+func (d *defaultDriver) scaleStatefulSetDown(
+	results *reconciler.Results,
+	statefulSet *appsv1.StatefulSet,
+	targetReplicas int,
+	esClient esclient.Client,
+	observedState observer.State,
+) error {
+	logger := log.WithValues("statefulset", k8s.ExtractNamespacedName(statefulSet))
+
+	if *statefulSet.Spec.Replicas == 0 && targetReplicas == 0 {
+		// we don't expect any new replicas in this statefulset, remove it
+		logger.Info("Deleting statefulset")
+		if err := d.Client.Delete(statefulSet); err != nil {
+			return err
+		}
+	}
+	// copy the current replicas, to be decremented with nodes to remove
+	updatedReplicas := int32(*statefulSet.Spec.Replicas)
+
+	// leaving nodes names can be built from StatefulSet name and ordinals
+	var leavingNodes []string
```

Comment:
a comment somewhere around here about the ordering/contents of leavingNodes would make sense. there seems to be an underlying assumption/invariant that it is in descending pod ordinal order. :)